tpu-provisioner/internal/cloud/gke.go

```diff
@@ -199,45 +213,80 @@ func (g *GKE) DeleteNodePool(name string, eventObj client.Object, why string) er
-		}
-		g.Recorder.Eventf(eventObj, corev1.EventTypeWarning, EventNodePoolDeletionFailed, "Request to delete Node Pool %s failed: %v.", name, err)
-		return fmt.Errorf("deleting node pool %q: %w", name, err)
-	}
-
-	if err := waitForGkeOp(g.Service, g.ClusterContext, op); err != nil {
-		g.Recorder.Eventf(eventObj, corev1.EventTypeWarning, EventNodePoolDeletionFailed, "Operation to delete Node Pool %s failed: %v.", name, err)
+		},
+		ReqFailure: func(err error) {
+			g.Recorder.Eventf(eventObj, corev1.EventTypeWarning, EventNodePoolDeletionFailed, "Request to delete Node Pool %s failed: %v.", name, err)
+		},
+		OpFailure: func(err error) {
+			g.Recorder.Eventf(eventObj, corev1.EventTypeWarning, EventNodePoolDeletionFailed, "Operation to delete Node Pool %s failed: %v.", name, err)
+		},
+		Success: func() {
+			g.Recorder.Eventf(eventObj, corev1.EventTypeNormal, EventNodePoolDeletionSucceeded, "Successfully deleted Node Pool %s.", name)
+		},
+	}); err != nil {
 		return err
 	}
 
-	g.Recorder.Eventf(eventObj, corev1.EventTypeNormal, EventNodePoolDeletionSucceeded, "Successfully deleted Node Pool %s.", name)
-
 	return nil
 }
 
 var ErrNodePoolStopping = errors.New("node pool stopping")
+var ErrNodePoolDeletedToBeRecreated = errors.New("node pool deleted to be recreated")
+
+type nodePoolState int
+
+const (
+	nodePoolStateUnknown nodePoolState = iota
+	nodePoolStateNotExists
+	nodePoolStateExistsAndMatches
+	nodePoolStateExistsAndNotMatches
+	nodePoolStateExistsAndStopping
+)
 
-func (g *GKE) nodePoolExists(name string) (bool, error) {
-	call := g.Service.Projects.Locations.Clusters.NodePools.Get(g.ClusterContext.NodePoolName(name))
-	np, err := call.Do()
+func (g *GKE) checkExistingNodePool(ctx context.Context, desired *containerv1beta1.NodePool) (nodePoolState, error) {
+	existing, err := g.NodePools.Get(ctx, desired.Name)
 	if err == nil {
-		return true, nil
+		match, err := nodePoolsMatch(desired, existing)
+		if err != nil {
+			return nodePoolStateUnknown, fmt.Errorf("comparing node pools: %w", err)
+		}
+		if match {
+			return nodePoolStateExistsAndMatches, nil
+		} else {
+			return nodePoolStateExistsAndNotMatches, nil
+		}
 	}
 	if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == http.StatusNotFound {
-		return false, nil
+		return nodePoolStateNotExists, nil
 	}
-	if np.Status == "STOPPING" {
-		return false, ErrNodePoolStopping
+	if existing.Status == "STOPPING" {
+		return nodePoolStateExistsAndStopping, nil
 	}
 
-	return false, err
+	return nodePoolStateUnknown, err
 }
 
-func (g *GKE) nodePoolForPod(name string, p *corev1.Pod) (*containerv1beta1.NodePool, error) {
+func nodePoolsMatch(desired, existing *containerv1beta1.NodePool) (bool, error) {
+	desiredHash, err := nodePoolSelectiveHash(desired)
+	if err != nil {
+		return false, fmt.Errorf("hashing node pool: %w", err)
+	}
+	if existing.Config != nil && existing.Config.Labels != nil {
+		existingHash, ok := existing.Config.Labels[LabelNodePoolHash]
+		if !ok {
```

Comment:
Would it make sense to make this behavior configurable? I imagine over time this case would happen less and less? 

I'd probably want to log a message as well to help with troubleshooting